impl/src/main/java/org/glassfish/soteria/servlet/AuthenticationData.java

```diff
@@ -40,7 +41,7 @@
     
     public AuthenticationData(Principal principal, Set<String> groups) {
         this.principal = principal;
-        this.groups =  unmodifiableSet(groups);
+        this.groups = new HashSet<>(unmodifiableSet(groups));
```

Comment:
What about either:

```java
new HashSet<>(groups);
```

or

```java
unmodifiableSet(new HashSet<>(groups));
```

The first makes a defensive copy, while the second makes a defensive copy and signals to the user it's useless to modify it.

Making a copy of an unmodifiable set doesn't seem to do anything ;)

Reply:
Ouch! It was too easy to make it wrong. ``unmodifiableSet(new HashSet<>(groups));` was really what I had in mind. I'll fix it.